daft/execution/physical_plan.py

```diff
@@ -263,43 +283,67 @@ def actor_pool_project(
+            for next_child in child_materializations:
+                if next_child.task.done():
+                    # Initialize actor pool upon first task that is ready to be submitted
+                    if actor_pool_id is None:
+                        actor_pool_id = actor_pool_manager.setup_actor_pool(
+                            actor_pool_name,
+                            actor_resource_request,
+                            task_resource_request,
+                            num_actors,
+                            projection,
+                        )
+
+                    actor_project_step = (
+                        PartitionTaskBuilder[PartitionT](
+                            inputs=[next_child.task.partition()],
+                            partial_metadatas=[next_child.task.partition_metadata()],
+                            actor_pool_id=actor_pool_id,
+                        )
+                        .add_instruction(
+                            instruction=execution_step.StatefulUDFProject(projection),
+                            resource_request=task_resource_request,
+                        )
+                        .finalize_partition_task_single_output(
+                            stage_id=stage_id,
+                        )
                     )
-                    .finalize_partition_task_single_output(
-                        stage_id=stage_id,
+                    actor_pool_materializations.append(IndexedPartitionTask(next_child.index, actor_project_step))
+                    yield actor_project_step
+                else:
+                    in_progress_child_materializations.append(next_child)
+
+            child_materializations = in_progress_child_materializations
+
+            # Pop ready actor_pool steps
+            in_progress_actor_pool_materializations = list()
+            for next_actor_pool_task in actor_pool_materializations:
+                if next_actor_pool_task.task.done():
+                    heapq.heappush(
+                        done_task_heap, IndexedPartitionTask(next_actor_pool_task.index, next_actor_pool_task.task)
                     )
-                )
-                actor_pool_materializations.append(actor_project_step)
-                yield actor_project_step
+                else:
+                    in_progress_actor_pool_materializations.append(next_actor_pool_task)
+
+            actor_pool_materializations = in_progress_actor_pool_materializations
 
-            # Exhaustively pop ready actor_pool steps and bubble it upwards as the start of a new pipeline
-            while len(actor_pool_materializations) > 0 and actor_pool_materializations[0].done():
-                next_ready_actor_pool_task = actor_pool_materializations.popleft()
+            # Bubble up completed tasks in order as the start of a new pipeline
+            while len(done_task_heap) > 0 and done_task_heap[0].index == next_expected_done_task_index:
```

Comment:
Could we build a little datastructure that can abstract away this `done_task_heap[0].index == next_expected_done_task_index` logic? It can keep track of the `next_expected_done_task_index` internally and return the next task only if it is available.